test/router_test.go

```diff
@@ -0,0 +1,216 @@
+	fillFrequency := 1000 / int(fillInterval.Milliseconds())
+	rate := 500
+	totalTxSent := 0
+
+	capacity := rate / fillFrequency
+	rl, err := armageddon.NewRateLimiter(rate, fillInterval, capacity)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to start a rate limiter")
+		os.Exit(3)
+	}
+
+	broadcastClient := client.NewBroadcastTxClient(uc, 10*time.Second)
+	startPullingChan := make(chan int, 1)
+
+	go func() {
+		wasRouterStopped := false
+		txSent := 0
+
+		for {
+			status := rl.GetToken()
+			if !status {
+				fmt.Fprintf(os.Stderr, "failed to send tx %d", txSent+1)
+				os.Exit(3)
+			}
+			txContent := armageddon.PrepareTx(txSent, 64, []byte("sessionNumber"))
+			err = broadcastClient.SendTx(txContent)
+			if err != nil {
+				// if a stream is closed, we assume the client is stopped
+				if strings.Contains(err.Error(), "rpc error: code = Internal desc = SendMsg called after CloseSend") ||
+					strings.Contains(err.Error(), "rpc error: code = Canceled desc = context canceled") ||
+					strings.Contains(err.Error(), "broadcast client is stopped") {
+					break
+				}
+
+				// EOF error is expected when a router is stopped
+				if strings.Contains(err.Error(), "EOF") {
+					startPullingChan <- txSent
+					require.False(t, wasRouterStopped)
+					wasRouterStopped = true
+				}
+				// we can continue sending transactions
+			} else if wasRouterStopped { // wait for a router to be restarted
+				startPullingChan <- txSent
+				wasRouterStopped = false
+			}
+
+			txSent++
+		}
+
+		startPullingChan <- txSent
+	}()
+
+	parties := []types.PartyID{}
+	for partyID := 1; partyID <= numOfParties; partyID++ {
+		parties = append(parties, types.PartyID(partyID))
+	}
+
+	time.Sleep(5 * time.Second)
+
+	// 7. Stop a router.
+	routerToStop := armaNetwork.GetRouter(t, types.PartyID(1))
+
+	t.Logf("Stopping router: party %d", routerToStop.PartyId)
+	routerToStop.StopArmaNode()
+
+	select {
+	case totalTxSent = <-startPullingChan:
+		t.Logf("Router %d stopped successfully", routerToStop.PartyId)
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "Router did not stop in time")
+	}
+
+	// 8. Pull blocks from all assemblers sent before a router was stopped.
+	PullFromAssemblers(t, &BlockPullerOptions{
+		UserConfig:       uc,
+		Parties:          parties,
+		StartBlock:       0,
+		EndBlock:         math.MaxUint64,
+		Transactions:     totalTxSent,
+		Timeout:          60,
+		NeedVerification: true,
+		ErrString:        "cancelled pull from assembler: %d",
+	})
+
+	time.Sleep(5 * time.Second)
+
+	// 9. Restart the router.
+	routerToStop.RestartArmaNode(t, readyChan, numOfParties)
+	testutil.WaitReady(t, readyChan, 1, 10)
+
+	select {
+	case totalTxSent = <-startPullingChan:
+		t.Logf("Router %d restarted successfully", routerToStop.PartyId)
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "Router did not restart in time")
+	}
+	correctParties := []types.PartyID{types.PartyID(2), types.PartyID(3), types.PartyID(4)}
```

Comment:
please pull from all assemblers and not only correct parties
all assemblers should behave the same